### DISCUSSION TITLE
Leaking client-components in multi-tenant websites

### DISCUSSION THREAD

**Discussion by jens-duttke** [vercel/next.js#58732] [Help]
### Summary

I have several customers whose websites with different domains are all based on predefined components. The most of the components can be easily shared between the different websites.
However, there are also customised components that are only used on specific websites.

I have now realised that the customer-specific (client-)components are also present in the JavaScript bundle of the other websites.
During my research, I realised that many Next.js pages have the same problem.

Even the example pages mentioned on https://demo.vercel.pub/platforms-starter-kit have this problem:

There is the "globe" icon in the social media bar on the https://akoskm.com/ website:
![image](https://github.com/vercel/next.js/assets/4883142/74ee5ce2-5001-4cf3-b7de-9b4b61e76434)

It is not present on https://eda.hashnode.dev/:
![image](https://github.com/vercel/next.js/assets/4883142/9a7f7102-21b0-42a5-846c-c32f94b1901a)

However, if you search for the SVG code here, you will find that it is nevertheless included in the JavaScript code:
![image](https://github.com/vercel/next.js/assets/4883142/fe1bff1a-a46d-4c8c-bcc4-adf8956ecb16)

The same with the "Newsletter", this is present on the first website, but not on the others, but the code is still there:
![image](https://github.com/vercel/next.js/assets/4883142/b8363242-be1c-476e-b5ef-9062862d830a)

This may not be a problem with such small, almost identical websites with a small number of components, but if I have relatively complex websites, like mine, with many customer-specific components, I see the problem that unnecessary code is transferred to the client, which causes unnecessary traffic and JavaScript processing on the client.

Pure client-side rendering with dynamically loaded components (similar to microfrontends) is out of the question for me, because the components should also be pre-rendered on the server side.

Because of the maintainability, I find the approach of using a Next.js server to deliver different pages via configuration very appealing. In addition, the hosting costs would be lower, as the individual pages only have relatively little traffic.

Does anyone have an idea how to prevent such leaking components?
Or is Next.js not suitable for larger multi-tenant sites?

**Comment by icyJoseph** [comment]
Do you have a file like:

```ts
// components/icons
export {Icon} from './Icon';
// and many more like this
```

So called barrel-file? Sometimes, the bundler cannot figure out which components to drop from the bundle, so it just bundles everything. 

With Next.js' dynamic function, you can lazily load components, and still have them in them be part of the SSR run. 

Do you have a link to a repository, or a code example we could talk over?

**Reply by jens-duttke** [reply]
I've just build a small, primitive example based on `create-next-app`:
https://github.com/jens-duttke/leaking-components

In the `src` folder, there is a folder "customer1" and a folder "customer2" (which represents components which shouldn't be loaded on each other website). This contains a server-component, a client-component and a async component loaded using next/dynamic.

In `src/app/[customer]/page.tsx` then either the components from "customer1" are rendered or the components from "customer2".

```
npm i
npm run build
npm run start
```

http://localhost:3000/customer1
http://localhost:3000/customer2

As you can see here, on the `customer1` page, the JavaScript code of the `customer2` is shown, and vice versa:

![image](https://github.com/vercel/next.js/assets/4883142/70696808-94e7-4daf-b4fb-ccef161136fb)
---
![image](https://github.com/vercel/next.js/assets/4883142/633cf4ce-3102-40d8-b6f0-6ceb8a219138)

**Reply by nest000** [reply]
anyone? I have exact the same issue!

**Reply by icyJoseph** [reply]
I gave this a go, and after fixing the hydration errors, I thought that doing this assignment

```js
const customers = { /* etc */ }
```

Was the issue, because, I'd assume the bundler, doesn't seem to have enough information to know that at runtime, we cannot just switch the customer parameter.

In other words, if we look at the code in isolation, forget Next.js, React, etc, we have two branches of code, controlled by a parameter, so of course when you call this function, you have to have the code for all paths.

- https://overreacted.io/how-does-the-development-mode-work/ - kind of what happens here, the bundler can eliminate certain paths... 

But changing to `const c1 = Customer1DynamicallyLoaded`, etc, and then even using a switch statement, to hint at the bundler that this code paths are clearly different, was not enough... the issue persisted... And sure, why would that give the bundler any more information about the distinct code paths.

So I thought, maybe if we use `generate static params`, then we might be able to label some code paths as unreachable, when each path gets built, but that was also not enough. I did not try using an `if (customers === 'something')` though, to try and trigger dropping the unused code, I had a switch case at that point.

I'd say, maybe, open an issue, and hopefully, someone from the core team can take a look and give a more detailed explanation, and ways forward.

**Reply by icyJoseph** [reply]
It still didn't make a difference to go with if statements:

```jsx
import dynamic from "next/dynamic";

const Customer1 = dynamic(
  () => import(/* webpackChunkName: "customer1" */ "../../../customer1")
);

const Customer2 = dynamic(
  () => import(/* webpackChunkName: "customer2" */ "../../../customer2")
);

export const generateStaticParams = () => {
  return ["customer1", "customer2"].map((customer) => ({ customer }));
};

export default function Page({ params }: { params: { customer: string } }) {
  if (params.customer === "customer1") {
    return <Customer1 />;
  }
  if (params.customer === "customer2") {
    return <Customer2 />;
  }
  return "not found";
}
```

The weird part is that it is a `page-something` file that contains this, the module itself for customer2 is never really loaded.
